enh: permit users with different versions of tonic

Right now tonic is used in two quite different ways: 1) it is an internal library that facilitates e.g. grpc handling and 2) tonic types are part of the public interface of opentelemetry-otlp.

For case (2) it would be really nice to have access to the version of tonic that opentelemetry-otlp expects as a re-export, rather than having to do lock-step upgrades of opentelemetry & tonic across our codebases.

something as simple as a pub use tonic in the root of opentelemetry-otlp might well be enough to fix this.

Example of (2) code today:
```
       TracingProtocol::Otlp(url, token) => {
            let mut map = tonic::metadata::map::MetadataMap::new();
            map.insert(ACCESS_TOKEN_HEADER, token.parse().unwrap());

            opentelemetry_otlp::new_pipeline()
                .tracing()
                .with_exporter(
                    opentelemetry_otlp::new_exporter()
                        .tonic()
                        .with_metadata(map)
                        .with_timeout(OTLP_TIMEOUT)
                        .with_protocol(Protocol::Grpc)
                        .with_endpoint(url),
                )
                .with_trace_config(trace_config)
                .install_batch(opentelemetry::runtime::Tokio)?
        }
```

The challenge here is that without reference to an exactly matching tonic version, this code will fail like so:
```
error[E0308]: mismatched types
Error:    --> src/tracing.rs:148:40
    |
148 |                         .with_metadata(map)
    |                          ------------- ^^^ expected struct `tonic::metadata::map::MetadataMap`, found struct `tonic::metadata::map::MetadataMap`
    |                          |
    |                          arguments to this function are incorrect
    |
    = note: perhaps two different versions of crate `tonic` are being used?
note: associated function defined here
   --> /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/opentelemetry-otlp-0.10.0/src/exporter/tonic.rs:44:12
    |
44  |     pub fn with_metadata(mut self, metadata: MetadataMap) -> Self {
    |            ^^^^^^^^^^^^^
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -102,7 +102,7 @@ jobs:
         run: |
           cargo install cargo-check-external-types@0.1.13
           cd ${{ matrix.example }}
-          cargo check-external-types --config allowed-external-types.toml
+          cargo check-external-types --all-features --config allowed-external-types.toml
   msrv:
     strategy:
       matrix:

--- a/opentelemetry-otlp/CHANGELOG.md
+++ b/opentelemetry-otlp/CHANGELOG.md
@@ -3,6 +3,8 @@
 ## vNext
 
 - Update `tonic` dependency version to 0.13
+- Re-export `tonic` types under `tonic_types`
+  [2898](https://github.com/open-telemetry/opentelemetry-rust/pull/2898)
 
 ## 0.29.0
 

--- a/opentelemetry-otlp/allowed-external-types.toml
+++ b/opentelemetry-otlp/allowed-external-types.toml
@@ -13,6 +13,7 @@ allowed_external_types = [
     # prost is a pre 1.0 crate
     "prost::error::EncodeError",
     # tonic is a pre 1.0 crate
+    "tonic",
     "tonic::status::Code",
     "tonic::status::Status",
     "tonic::metadata::map::MetadataMap",

--- a/opentelemetry-otlp/allowed-external-types.toml
+++ b/opentelemetry-otlp/allowed-external-types.toml
@@ -3,23 +3,16 @@
 # This is used with cargo-check-external-types to reduce the surface area of downstream crates from
 # the public API. Ideally this can have a few exceptions as possible.
 allowed_external_types = [
-    "opentelemetry::*",
     "opentelemetry_http::*",
     "opentelemetry_sdk::*",
-    # http is a pre 1.0 crate
-    "http::uri::InvalidUri",
-    "http::header::name::InvalidHeaderName",
-    "http::header::value::InvalidHeaderValue",
-    # prost is a pre 1.0 crate
-    "prost::error::EncodeError",
+    # serde
+    "serde::de::Deserialize",
+    "serde::ser::Serialize",
     # tonic is a pre 1.0 crate
-    "tonic::status::Code",
-    "tonic::status::Status",
     "tonic::metadata::map::MetadataMap",
     "tonic::transport::channel::tls::ClientTlsConfig",
     "tonic::transport::tls::Certificate",
     "tonic::transport::tls::Identity",
     "tonic::transport::channel::Channel",
-    "tonic::transport::error::Error",
     "tonic::service::interceptor::Interceptor",
 ]

--- a/opentelemetry-otlp/allowed-external-types.toml
+++ b/opentelemetry-otlp/allowed-external-types.toml
@@ -13,10 +13,12 @@ allowed_external_types = [
     # prost is a pre 1.0 crate
     "prost::error::EncodeError",
     # tonic is a pre 1.0 crate
-    "tonic",
     "tonic::status::Code",
     "tonic::status::Status",
     "tonic::metadata::map::MetadataMap",
+    "tonic::transport::channel::tls::ClientTlsConfig",
+    "tonic::transport::tls::Certificate",
+    "tonic::transport::tls::Identity",
     "tonic::transport::channel::Channel",
     "tonic::transport::error::Error",
     "tonic::service::interceptor::Interceptor",

--- a/opentelemetry-otlp/src/lib.rs
+++ b/opentelemetry-otlp/src/lib.rs
@@ -451,13 +451,16 @@ pub enum Protocol {
 /// Placeholder type when no exporter pipeline has been configured in telemetry pipeline.
 pub struct NoExporterConfig(());
 
+/// Re-exported types from the `tonic` crate.
 #[cfg(feature = "grpc-tonic")]
 pub mod tonic_types {
+    /// Re-exported types from `tonic::metadata`.
     pub mod metadata {
         #[doc(no_inline)]
         pub use tonic::metadata::MetadataMap;
     }
 
+    /// Re-exported types from `tonic::transport`.
     #[cfg(feature = "tls")]
     pub mod transport {
         #[doc(no_inline)]

--- a/opentelemetry-otlp/src/lib.rs
+++ b/opentelemetry-otlp/src/lib.rs
@@ -407,3 +407,7 @@ pub enum Protocol {
 #[doc(hidden)]
 /// Placeholder type when no exporter pipeline has been configured in telemetry pipeline.
 pub struct NoExporterConfig(());
+
+#[cfg(feature = "grpc-tonic")]
+#[doc(no_inline)]
+pub use tonic;

--- a/opentelemetry-otlp/src/lib.rs
+++ b/opentelemetry-otlp/src/lib.rs
@@ -452,5 +452,15 @@ pub enum Protocol {
 pub struct NoExporterConfig(());
 
 #[cfg(feature = "grpc-tonic")]
-#[doc(no_inline)]
-pub use tonic;
+pub mod tonic_types {
+    pub mod metadata {
+        #[doc(no_inline)]
+        pub use tonic::metadata::MetadataMap;
+    }
+
+    #[cfg(feature = "tls")]
+    pub mod transport {
+        #[doc(no_inline)]
+        pub use tonic::transport::{Certificate, ClientTlsConfig, Identity};
+    }
+}

